### ISSUE TITLE
StateAdapt + Angular: Easy way to share state?

### ISSUE THREAD

**Issue by diegovilar** [state-adapt/state-adapt#44]
Congrats on the work with StateAdapt, @mfp22. The boilerplate of almost every other state management library always throws me off. The declarative and simplistic approach you implemented is very welcome and I hope it gets used more and more. 👏👏

I've been experimenting with a few libraries in search of one that is easy enough to use in my workplace. I've been using other libraries for a long time, but the whole reactive paradigm is already alien enough for my colleagues that come from a Java back-end background, so I'm trying to define an architecture and stack in which they can be productive without having to give them a very long lecture on front-end development, Angular and specially state management boilerplate. As far as state management libraries come, SA is looking pretty good so far :)

While trying some scenarios while evaluating it, I have not found a recommended way to share state throughout the application: a global state, per se. Of course the whole state is global, but I'm talking about a shared access to what you called "path", which is what I understand to be an slice of the global state. I'm also interesting on being able to not only watch it, so `watch()` is not enough.

I know that if I take control of the creation of the store instead of relying on `@state-adapt/angular` utilities I can accomplish this, but then I'd be giving up on it's `adapt()` and `watch()` utilities and angular's DI they rely on.

I cannot use `adapt("<some-path>")` in multiple components because, as you clearly explain in the docs, that would throw since I'd actually be trying to create another "adapted state" with the same path.

Is there an easy way to "define" a shared adapted state to be used in multiple places? I also know that could be done with a service, but I'm perhaps looking for something more "state-like", 

While experimenting, I came up with a `defineState()` utility that can be used like this:

```typescript
// user-state.ts

import { defineState } from "@experiments/state-adapt";

export const USER_STATE_PATH = "user";

export type User = Readonly<{
  name: string;
  email: string;
  // ...
}>;

const INITIAL_STATE: User = {
  name: "Diego",
  email: "xxxx@gmail.com",  
  // ...
};

export const [useUserState, watchUserState] = defineState(USER_STATE_PATH, INITIAL_STATE);

// defineState() has the same signature as adapt(), but returns a tuple with delegates
// to adapt() and watch(). A closure keeps track of the return of adapt() so useUserState()
// is memoized.
```

Later on, in componentes, services, etc:

```typescript
// sample.component.ts

import { adapt } from '@state-adapt/angular';
import { useUserState } from "@state/user-state";

@Component({/* ... */})
export SampleComponent {
  protected readonly userState = useUserState(); // shared state
  protected readonly state = adapt(/* ... */);   // component's local state
  // ...
}
```

```typescript
// other-sample.component.ts

import { useUserState } from "@state/user-state";

@Component({/* ... */})
export OtherSampleComponent {
  protected readonly userState = useUserState(); // shared state
  // ...
}
```

```typescript
// another-sample.component.ts

import { watchUserState } from "@state/user-state";

@Component({/* ... */})
export AnotherSampleComponent {
  protected readonly userState = watchUserState(/* custom selectors */); // shared state
  // ...
}
```

Is there a better way to do this? I'm not sure if I'm missing something in the docs or if I'm trying to use the library in a way it was not intended to be used.

As I said before, I know that I could just create a service in which I would use `adapt(...)` and expose it directly or through a facade, but to introduce a thin service as an indirection seems like boilerplate too and I'm trying to keep it as simple as possible for my colleagues.

**Comment by mfp22** [comment]
> and I hope it gets used more and more. 👏👏

Thank you!

> the whole reactive paradigm is already alien enough for my colleagues that come from a Java back-end background
> be productive without having to give them a very long lecture on front-end development

That sounds hard to do. Reactive programming will be counterintuitive for them at first. 

> As far as state management libraries come, SA is looking pretty good so far :)

That's good. I was sad about how difficult reactive programming is for everyone and my goal was to try to make it as easy as possible by getting rid of the extra stuff. But Java devs are used to boilerplate, so I think the hardest part will be the "thinking reactively" part. 

I'm starting to think about creating a StateAdapt tutorial. I honestly don't think I'm the best person to do this. I'm better at talking about weird stuff than helping new people. But there is a need. 

But if I were you I would create 3 StackBlitz exercises for them. 1 for each mindset shift required. Here are 3 mindset shifts: https://twitter.com/mfpears/status/1680469781290549249

The solutions are always to avoid writing methods. For 1, use a `computed` signal. For 2, use a `Source` and have 2 stores react. For 3, have them make a typeahead input. The solution is to start with a Subject or Source and define `results$ = search$.pipe(...)`. 

You could find existing StackBlitz solutions and take away the solution let them practice solving it reactively instead of imperatively, so they have to think about it.

---

> I have not found a recommended way to share state throughout the application

Just use a service. That's the official answer for all Angular stuff. There are other ways, but this will be the path of least resistance. 

```tsx
@Injectable({ providedIn: 'root' })
export class UserStore {
  store = adapt(['user', {} as User], {});
}

// ...

@Component({...})
export class UserComponent {
  user = inject(UserStore).store;
}
```

Below you say,

> I also know that could be done with a service, but I'm perhaps looking for something more "state-like",

I think you might be referring to React hooks. But I assure you that you will have the easiest time if you share state in Angular using services. 

If you want an alternative syntax to class services, I've done this before:

```typescript
export const UserStore = new InjectionToken('UserStore', {
  providedIn: 'root',
  factory: () => {
    // Set up code, or just return 
    return adapt(['user', {} as User], adapter)
  },
});

// ...

@Component({...})
export class UserComponent {
  userStore = inject(UserStore); // No need to access .store
}
```

I would caution using that syntax though, because it is uncommon and if you move state from a component you will need to adjust the syntax (add `const` and remove `this.` everywhere).

---

However... If I wanted to I could export a function from StateAdapt like this:

```typescript
export injectUser = createInjectAdapt(['user', {} as User, userAdapter], () => {
  // I think only sources need access to DI... NgRx has gone years without reducers having it
  user$ = inject(HttpClient).get('/user');
  return user$; // Or full Sources object https://state-adapt.github.io/angular/docs/angular#overload-3
));
```

It would reuse the `user` state path as the injection token. This is sounding really good...

It would also solve the problem of going from component to shared, because within that factory function you could use `const` instead of `this`. 

Oh, and I could even pass a detached store (result of `watch`) to it, for recursive stuff without having to explicitly use `watch`: 

```typescript
export const injectCount = createInjectAdapt(
  ['count', 0],
  watched => watched.state$.pipe(
    delay(1000),
    map(n => n + 1),
    toSource('tick'),
  ),
);
```

This is really interesting, but I need to think a lot more about it, and it would either eventually create another overload for `adapt`, or make it a breaking change, so it would have to wait for version 2 possibly. And there are other non-breaking changes I need to make first. 

So, I would definitely say for now just use a regular service. It's by far the most common way to share state in Angular.

**Comment by mfp22** [comment]
I got this to work without a breaking change. I just added the option to define `sources` as a function taking in a store, and all the types are working nicely, and it's functioning correctly. It was actually very easy to do. I need to add some tests and update docs, but I could even release this in a version within the next week.

But overall it doesn't appear to actually reduce boilerplate very much. And it adds complexity to the API and documentation. I think it really might be better to just go with the common Angular service class approach here. But I will sit on this a bit longer and decide what to do later.

**Comment by mfp22** [comment]
This has now been released in version 1.2.0

I named the function `adaptInjectable`, and I would advise caution before using it. It's "experimental" for now. I released it so I could easily test it out in various projects, and maybe other people can too and let me know what they think. 

For now you can see how it changed one codebase here https://github.com/mfp22/2023-fidelity-workshop/commit/33d7311944263802953c294a31879eaae6d6f5f3

It didn't reduce lines of code, but it did slim down some code a little bit. 

It's not providing much value so far, so most likely won't be worth maintaining in the long run. Regular services are fine I think. But at least it's there to try out.
